corehq/apps/locations/fixtures.py

```diff
@@ -70,47 +70,87 @@ def __call__(self, user, version, last_sync=None, app=None):
         """
         if not user.project.uses_locations:
             return []
-        if toggles.SYNC_ALL_LOCATIONS.enabled(user.domain):
-            locations = SQLLocation.active_objects.filter(domain=user.domain)
-        else:
-            locations = []
-            user_location = user.sql_location
-            if user_location:
-                # add users location (and ancestors) to fixture
-                locations.append(user_location)
-
-                # add all descendants as well
-                locations += (user_location.get_descendants()
-                                           .filter(is_archived=False))
-
-            if user.project.supports_multiple_locations_per_user:
-                # this might add duplicate locations but we filter that out later
-                location_ids = [loc.location_id for loc in user.locations]
-                locations += SQLLocation.active_objects.filter(
-                    location_id__in=location_ids
-                )
-
-        location_db = _location_footprint(locations)
-
-        if not should_sync_locations(last_sync, location_db, user):
-            return []
 
-        root = Element('fixture',
-                       {'id': self.id,
-                        'user_id': user.user_id})
+        all_locations = _all_locations(user)
+
+        if not should_sync_locations(last_sync, all_locations, user):
+            return []
 
+        root_node = Element('fixture', {'id': self.id, 'user_id': user.user_id})
         root_locations = filter(
-            lambda loc: loc.parent is None, location_db.by_id.values()
+            lambda loc: loc.parent is None, all_locations.by_id.values()
         )
 
         if root_locations:
-            _append_children(root, location_db, root_locations)
-        return [root]
+            _append_children(root_node, all_locations, root_locations)
+        return [root_node]
 
 
 location_fixture_generator = LocationFixtureProvider()
 
 
+def _all_locations(user):
+    if toggles.SYNC_ALL_LOCATIONS.enabled(user.domain):
+        return LocationSet(SQLLocation.active_objects.filter(domain=user.domain))
+    else:
+        leaf_locations = {leaf_location for leaf_location in _gather_leaf_locations(user)}
+        return _location_footprint(leaf_locations)
```

Comment:
Did you look in to dropping `_location_footprint` at all?  I imagine that's a fair performance hog for domains with a lot of locations, as it looks like it makes a LOT of db calls.

I'd expect it to be possible to just get the full list of locations directly, rather than walking the tree.  Something along the lines of:

```python
ancestors = location_to_expand_from.get_ancestors(include_self=True)
descendants = location_to_expand_from.get_descendants().filter(location_type__in=location_types_to_sync)
return ancestors + descendants
```

I haven't actually tried it out though, so it's entirely possible I'm missing something.